rust/blockstore/src/arrow/ordered_blockfile_writer.rs

```diff
@@ -0,0 +1,956 @@
+use uuid::Uuid;
+
+// The end key is exclusive, if the end key is None, then the block/delta is open-ended
+type BlockIdAndEndKey = (Uuid, Option<CompositeKey>);
+type CurrentDeltaAndEndKey = (OrderedMergeBlockDelta, Option<CompositeKey>);
+
+#[derive(Clone)]
+pub struct ArrowOrderedBlockfileWriter {
+    block_manager: BlockManager,
+    root_manager: RootManager,
+    remaining_block_stack: Arc<Mutex<VecDeque<BlockIdAndEndKey>>>, // todo: rename
+    current_block_delta: Arc<Mutex<Option<CurrentDeltaAndEndKey>>>,
+    completed_block_deltas: Arc<Mutex<Vec<OrderedMergeBlockDelta>>>,
+    root: RootWriter,
+    id: Uuid,
+}
+// TODO: method visibility should not be pub(crate)
+
+#[derive(Error, Debug)]
+pub enum ArrowBlockfileError {
+    #[error("Block not found")]
+    BlockNotFound,
+}
+
+impl ChromaError for ArrowBlockfileError {
+    fn code(&self) -> ErrorCodes {
+        match self {
+            ArrowBlockfileError::BlockNotFound => ErrorCodes::Internal,
+        }
+    }
+}
+
+impl ArrowOrderedBlockfileWriter {
+    pub(super) fn new<K: ArrowWriteableKey, V: ArrowWriteableValue>(
+        id: Uuid,
+        block_manager: BlockManager,
+        root_manager: RootManager,
+    ) -> Self {
+        let initial_block = block_manager.create::<K, V, OrderedMergeBlockDelta>();
+        let sparse_index = SparseIndexWriter::new(initial_block.id);
+        let root_writer = RootWriter::new(CURRENT_VERSION, id, sparse_index);
+
+        Self {
+            block_manager,
+            root_manager,
+            current_block_delta: Arc::new(Mutex::new(Some((initial_block, None)))),
+            completed_block_deltas: Arc::new(Mutex::new(Vec::new())),
+            root: root_writer,
+            remaining_block_stack: Arc::new(Mutex::new(VecDeque::new())), // todo: how to make sure this isn't concurrently modified
+            id,
+        }
+    }
+
+    pub(super) fn from_root(
+        id: Uuid,
+        block_manager: BlockManager,
+        root_manager: RootManager,
+        new_root: RootWriter,
+    ) -> Self {
+        let existing_block_deliminators = {
+            let root_forward = &new_root.sparse_index.data.lock().forward;
+
+            root_forward
+                .iter()
+                .zip_longest(root_forward.iter().skip(1))
+                .map(|zipped| match zipped {
+                    itertools::EitherOrBoth::Both(block, next_block) => match next_block.0 {
+                        SparseIndexDelimiter::Start => {
+                            panic!("Invariant violated: unexpected start delimiter.")
+                        }
+                        SparseIndexDelimiter::Key(end_key) => (*block.1, Some(end_key.clone())),
+                    },
+                    itertools::EitherOrBoth::Left(block) => (*block.1, None),
+                    itertools::EitherOrBoth::Right(_) => {
+                        unreachable!()
+                    }
+                })
+                .collect::<VecDeque<_>>()
+        };
```

Comment:
discussion: since we take a snapshot of the current blocks here, how do we guarantee that the sparse index is not updated by something else? is this already handled because we don't expose the sparse index outside of the writer and there can only be one writer?

Reply:
I think its because there is only one writer so the data hierarchy is BlockfileWriter -> Sparse Index. I.E ensuring these are in sync is writers job